File: scala/73.scala
import scala.collection.mutable
object Solution {
    def setZeroes(matrix: Array[Array[Int]]): Unit = {
        def set(i: Int, j: Int) {
            var _i = i-1
            var _j = j-1
            while (_i >= 0 && matrix(_i)(j) != 0) matrix(_i)(j) = 0
            _i = i+1
            while (_i < matrix.length && matrix(_i)(j) != 0) matrix(_i)(j) = 0
            while (_j >= 0 && matrix(i)(_j) != 0) matrix(i)(_j) = 0
            _j = j+1
            while (_j < matrix.head.length && matrix(i)(_j) != 0) matrix(i)(_j) = 0
        } 
        val s = mutable.Set[IndexedSeq[Int]]()
        for (i <- 0 until matrix.length; j <- 0 until matrix.head.length) 
            s.add((i, j))
        
        for (e <- s)
            set(e(0), e(1))
    }
}

File: leetcodeJava/73.java
package leetcodeJava;

class Solution {
    public void setZeroes(int[][] matrix) {
        boolean r = false, c = false;
        int m = matrix.length, n = matrix[0].length;
        for (var i = 0; i < m; i++)
            if (matrix[i][0] == 0) c = true;
        for (var j = 0; j < n; j++)
            if (matrix[0][j] == 0) r = true;
        
        for (var i = 0; i < m; i++) {
            for (var j = 0; j < n; j++) {
                if (matrix[i][j] == 0) {
                    if (i == 0) r = true;
                    if (j == 0) c = true;
                    matrix[i][0] = 0;
                    matrix[0][j] = 0;
                }
            }
        }
        for (var i = 1; i < m; i++) {
            for (var j = 1; j < n; j++) {
                if (matrix[i][0] == 0 || matrix[0][j] == 0) {
                    matrix[i][j] = 0;
                }
            }
        }
        if (r) 
            for (var j = 0; j < n; j++) matrix[0][j] = 0;
        if (c)
            for (var i = 0; i < m; i++) matrix[i][0] = 0;
    }
}

